gdnative-core/src/geom/basis.rs

```diff
@@ -29,14 +116,316 @@ impl Basis {
+    }
+
+    /// Returns the inverse of the matrix.
+    pub fn inverted(mut self) -> Basis {
+        self.invert();
+        self
+    }
+
+    /// Transposes the matrix.
+    pub fn transpose(&mut self) {
+        std::mem::swap(&mut self.elements[0].y, &mut self.elements[1].x);
+        std::mem::swap(&mut self.elements[0].z, &mut self.elements[2].x);
+        std::mem::swap(&mut self.elements[1].z, &mut self.elements[2].y);
+    }
+
+    /// Returns the transposed version of the matrix.
+    pub fn transposed(mut self) -> Basis {
+        self.transpose();
+        self
+    }
+
+    /// Returns the determinant of the matrix.
+    pub fn determinant(&self) -> f32 {
+        let elements = self.elements;
+        return elements[0].x * (elements[1].y * elements[2].z - elements[2].y * elements[1].z)
+            - elements[1].x * (elements[0].y * elements[2].z - elements[2].y * elements[0].z)
+            + elements[2].x * (elements[0].y * elements[1].z - elements[1].y * elements[0].z);
+    }
+
+    /// Orthonormalizes the matrix
+    ///
+    /// Performs a [Gram-Schmidt orthonormalization](https://en.wikipedia.org/wiki/Gram-Schmidt_process) on the basis of the matrix.
+    /// This can be useful to call from time to time to avoid rounding error for orthogonal matrices.
+    pub fn orthonormalize(&mut self) {
+        assert!(
+            !self.determinant().approx_eq(&0.0),
+            "Determinant should not be zero."
+        );
+
+        // Gram-Schmidt Process
+        let mut x = self.x();
+        let mut y = self.y();
+        let mut z = self.z();
+
+        x = x.normalize();
+        y = y - x * (x.dot(y));
+        y = y.normalize();
+        z = z - x * (x.dot(z)) - y * (y.dot(z));
+        z = z.normalize();
+
+        self.set_x(x);
+        self.set_y(y);
+        self.set_z(z);
+    }
+
+    /// Returns an orthonormalized version of the matrix.
+    ///
+    /// See [`Basis::orthonormalize()`](./struct.Basis.html#method.orthonormalize)
+    pub fn orthonormalized(mut self) -> Basis {
+        self.orthonormalize();
+        self
+    }
+
+    /// Returns true if this basis and b are approximately equal.
+    pub fn approx_eq(&self, other: &Basis) -> bool {
+        return self.elements[0].approx_eq(&other.elements[0])
+            && self.elements[1].approx_eq(&other.elements[1])
+            && self.elements[2].approx_eq(&other.elements[2]);
+    }
+
+    fn is_orthogonal(&self) -> bool {
+        let identity = Self::identity();
+        let m = (*self) * self.transposed();
+        m.approx_eq(&identity)
+    }
+
+    fn is_rotation(&self) -> bool {
+        let det = self.determinant();
+        det.approx_eq(&1.0) && self.is_orthogonal()
+    }
+
+    /// Multiplies the matrix from left by the rotation matrix: M -> R.M
+    ///
+    /// The main use of Basis is as Transform.basis, which is used a the transformation matrix
+    /// of 3D object. Rotate here refers to rotation of the object (which is R * (*this)),
```

Comment:
Add `` ` `` around identifiers and code. Translate cpp-isms to Rust syntax and names.